### PR TITLE
Faster wires

### PR DESCRIPTION
**Pull request by ziofil**
**Context:**
New, faster wires, and a few touches here and there.

**Description of the Change:**
Wires now works with sets and it's unordered in its internal operation. When it has to return modes in an order it will choose increasing order. 

**Benefits:**
Much simpler implementation, super fast

### CODE REVIEW

File: mrmustard/lab_dev/states/base.py [hunk 1]
```diff
@@ -82,7 +80,7 @@ class Ket(State):
     """
 
     def __init__(self, name: Optional[str] = None, modes: Optional[Sequence[int]] = None):
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:82]:
```suggestion
    def __init__(self, name: str = "Ket", modes: tuple[int,...] = ()):
```

File: mrmustard/lab_dev/states/base.py [hunk 2]
```diff
@@ -46,9 +46,7 @@ class DM(State):
         modes: The modes of this state.
     """
 
-    def __init__(self, name: Optional[str] = None, modes: Optional[Sequence[int]] = None):
-        modes = modes or []
-        name = name or ""
+    def __init__(self, name: str = "DM", modes: tuple[int,...] = ()):
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:49]: I think you need to modify Unitary and Channel in the same way

File: mrmustard/lab_dev/states/states.py [hunk 3]
```diff
@@ -115,7 +115,7 @@ class Number(Ket):
     """
 
     def __init__(
-        self, modes: Iterable[int], n: Union[int, Iterable[int]], cutoff: Optional[int] = None
+        self, modes: Sequence[int], n: Union[int, Sequence[int]], cutoff: Optional[int] = None
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/states.py:118]: DM and Ket have `modes = tuple[int, ...]`, wires has sets, here there are sequences. Choose one?

File: mrmustard/lab_dev/wires.py [hunk 4]
```diff
@@ -15,422 +15,211 @@
...
-# pylint: disable=protected-access
 class Wires:
     r"""
     A class with wire functionality for tensor network applications.
 
-    In MrMustard, we represent circuit components as tensors in a tensor network. The wires of
-    these components describe how they connect with the surrounding components. For example, an
-    `N`-mode pure state has `N` ket wires on the output side, while a `N`-mode mixed state
-    has `N` ket wires and `N` bra wires on the output side.
+    In MrMustard, instances of `CircuitComponent` have a ``Wires`` attribute.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/wires.py:27]:
```suggestion
    In MrMustard, instances of ``CircuitComponent`` have a ``Wires`` attribute.
```

File: mrmustard/lab_dev/states/states.py (hunk 3 above)
Comment (ziofil) [reply mrmustard/lab_dev/states/states.py:118]: Let's use tuples throughout (small, immutable). Only Wires will have sets. I'm not going to chase after all the type annotations in this PR though.

**Review by SamFerracin — COMMENTED** [review]
Looks very good! 
I left a few minor comments. Other than that, there are a few things to discuss, namely:
- Allowed types for `modes` for the various classes
- What to do with the parameter set (which we may want to "fix" in another PR though, seeing that it's an issue that exists previous to this PR?)
Good to merge after that discussion :)

File: mrmustard/lab_dev/wires.py [hunk 5]
```diff
@@ -15,422 +15,255 @@
...
-                rows_html[row_idx] += cell_html
-
-        # Close the rows and add them to the HTML table
-        for row_html in rows_html:
-            row_html += "</tr>"
-            html += row_html
-
-        html += "</table>"
-        display(HTML(html))
+        return f"Wires{self.args}"
```
Comment (SamFerracin) [inline mrmustard/lab_dev/wires.py:269]: https://github.com/XanaduAI/MrMustard/pull/361#discussion_r1518396850

Comment (ziofil) [reply mrmustard/lab_dev/wires.py:338]: not sure I follow 😅

File: tests/test_lab_dev/test_wires.py [hunk 6]
```diff
@@ -27,154 +26,74 @@ class TestWires:
...
-
-    def test_ids(self):
-        w = Wires([0, 1], [2], [3, 4, 5], [6])
-
-        assert w.output.bra.ids == w.ids[:2]
-        assert w.input.bra.ids == [w.ids[2]]
-        assert w.output.ket.ids == w.ids[3:6]
-        assert w.input.ket.ids == [w.ids[-1]]
-
-    def test_ids_setter(self):
```
Comment (SamFerracin) [inline tests/test_lab_dev/test_wires.py:98]: @ziofil Could you write a test to show how one can set the ids?

- Why it's useful to set the `id`s:
To connect 2 components while constructing the graph, which is useful to either validate the user-provided path (simulator) or to find the optimal path

- How we do this now:
By manually changing the ids of the wires that contract

File: mrmustard/lab_dev/circuit_components.py [hunk 7]
```diff
@@ -48,16 +48,35 @@ class CircuitComponent:
...
+        if a != sorted(a) or b != sorted(b) or c != sorted(c) or d != sorted(d):
+            offsets = [0, len(a), len(a) + len(b), len(a) + len(b) + len(c)]
+            perm = (
+                tuple(np.argsort(a))
+                + tuple(np.argsort(b) + offsets[0])
+                + tuple(np.argsort(c) + offsets[1])
+                + tuple(np.argsort(d) + offsets[2])
+            )
+            if self._representation is not None:
+                self._representation = self._representation.reorder(tuple(perm))
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components.py:79]: :+1:  pretty cool
Just a reminder that we will need to write the same piece of code to reorder the parameter set

File: mrmustard/lab_dev/states/base.py [hunk 8]
```diff
@@ -65,6 +65,7 @@ def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
             dm = DM()
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:65]: I don't know why we're not using `_from_attributes` here (I probably forgot to update this), but I believe that `dm = DM._from_attributes("", component.representation, component.wires)` should do the job of these 4 lines

File: mrmustard/lab_dev/states/base.py [hunk 9]
```diff
@@ -101,11 +102,13 @@ def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
             ket = Ket()
             ket._wires = component.wires
             ket._representation = component.representation
+            ket._name += "".join(str(m) for m in component.modes)
             return ket
         elif isinstance(other, Channel) and set(other.modes).issubset(set(self.modes)):
             dm = DM()
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:108]: ``dm = DM._from_attributes("", component.representation, component.wires)``

File: mrmustard/lab_dev/transformations/base.py [hunk 10]
```diff
@@ -65,11 +65,13 @@ def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
             unitary = Unitary()
```
Comment (SamFerracin) [inline mrmustard/lab_dev/transformations/base.py:65]: ``unitary = Unitary._from_attributes("", component.representation, component.wires)``

File: mrmustard/lab_dev/transformations/base.py [hunk 11]
```diff
@@ -65,11 +65,13 @@ def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
...
             unitary._wires = component.wires
             unitary._representation = component.representation
+            unitary._name += "".join(str(m) for m in component.modes)
             return unitary
         elif isinstance(other, Channel):
             channel = Channel()
```
Comment (SamFerracin) [inline mrmustard/lab_dev/transformations/base.py:71]: ``channel = Channel._from_attributes("", component.representation, component.wires)``

File: mrmustard/lab_dev/transformations/base.py [hunk 12]
```diff
@@ -107,6 +111,7 @@ def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
...
             channel._wires = component.wires
             channel._representation = component.representation
+            channel._name += "".join(str(m) for m in component.modes)
```
Comment (SamFerracin) [inline mrmustard/lab_dev/transformations/base.py:114]: ``channel = Channel._from_attributes("", component.representation, component.wires)``

File: mrmustard/lab_dev/states/states.py [hunk 13]
```diff
@@ -33,7 +33,7 @@
 class Coherent(Ket):
     r"""The `N`-mode coherent state.
 
-    If ``x`` and/or ``y`` are iterables, their length must be equal to `1` or `N`. If their length is equal to `1`,
+    If ``x`` and/or ``y`` are Sequences, their length must be equal to `1` or `N`. If their length is equal to `1`,
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/states.py:36]:
```suggestion
    If ``x`` and/or ``y`` are ``Sequence``\s, their length must be equal to `1` or `N`. If their length is equal to `1`,
```

File: mrmustard/lab_dev/states/states.py [hunk 14]
```diff
@@ -165,7 +165,7 @@ class Vacuum(Ket):
 
     def __init__(
         self,
-        modes: Iterable[int],
+        modes: tuple[int, ...],
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/states.py:168]:
```suggestion
        modes: Sequence[int],
```

**Review by SamFerracin — COMMENTED** [review]
Great job!
Most of the comments are about docs, or things that I forgot in previous PRs :sweat_smile: 

The biggest question I have is related to the `id`s: how can we set them, when we want to connect 2 components? See the comment in `test_wires.py`

Can you also pull develop?

File: mrmustard/lab_dev/wires.py [hunk 15]
```diff
@@ -15,422 +15,342 @@
...
+
+    @cached_property
+    def sorted_args(self) -> tuple[list[int], ...]:
+        r"The sorted arguments. Allows to sort them only once."
+        return tuple(sorted(s) for s in self.args)
+
+    @cached_property
+    def original(self):
+        r"The 'parent' ``Wires`` object, if any."
+        return self._original or self
```
[the remark this answers is not in the document]
Comment (ziofil) [reply mrmustard/lab_dev/wires.py:201]: nope

**Comment by SamFerracin** [comment]
I fixed the tests and addressed the last comments.
However, I wasn't able to address the most important one, namely how to change the ids ([this comment](https://github.com/XanaduAI/MrMustard/pull/362#discussion_r1528624971)). I am worried that they may not be changeable because of the caching. Let's make sure before we merge

**Review by SamFerracin — APPROVED** [review]
:rocket:

**Pull request merged**